accepted/2023/preview-apis/preview-apis.md

```diff
@@ -0,0 +1,194 @@
+viral nature is also undesirable.
+
+## Scenarios and User Experience
+
+## Requirements
+
+### Goals
+
+* Don't change the guardrails for platform-level preview features:
+    - They require opt-in (without, the build fails with an error)
+    - They require to be viral (i.e. opt-in is also required for indirect
+      dependencies relying on preview features)
+    - The opt-in can be expressed centrally in the project file
+    - They require SDK and framework to match
+
+* Offer library-level preview features with a different set of guardrails:
+    - They require opt-in (without, the build fails with an error)
+    - The opt-in isn't viral and can be an implementation detail
+    - The opt-in is centralized, but it's specific to each preview feature
+    - Using them poses no requirements on the SDK version
+
+### Non-Goals
+
+* Relaxing or extending the guardrails for platform-level preview features
+
+## Stakeholders and Reviewers
+
+* Runtime teams
+* C#, VB, F# compiler and IDE teams
+* Library teams
+    - Core libraries
+    - ASP.NET
+    - EF
+    - Azure SDK
+    - R9
+* Project system team
+* SDK team
+* MSBuild team
+* NuGet team
+
+## Design
+
+### Attribute
+
+We'll create a new attribute, separate from the existing `System.Runtime.Versioning.RequiresPreviewFeaturesAttribute`:
+
+```C#
+namespace System.Diagnostics.CodeAnalysis;
+
+[AttributeUsage(AttributeTargets.Assembly |
+                AttributeTargets.Module |
+                AttributeTargets.Class |
+                AttributeTargets.Struct |
+                AttributeTargets.Enum |
+                AttributeTargets.Constructor |
+                AttributeTargets.Method |
+                AttributeTargets.Property |
+                AttributeTargets.Field |
+                AttributeTargets.Event |
+                AttributeTargets.Interface |
+                AttributeTargets.Delegate, Inherited = false)]
+public sealed class PreviewAttribute : Attribute
+{
+    public PreviewAttribute(string diagnosticId);
+    public string DiagnosticId { get; }
+    public string? UrlFormat { get; set; }
+}
+```
+
+### Compiler Behavior
+
+The compiler will raise a diagnostic when a preview API is used, using supplied
+diagnostic ID. The severity is always error.
+
+The semantics are identical to how obsolete is tracked, except that it will
+always raise a diagnostic regardless of whether the call site is marked as
+preview or not. There is also no special treatment when both caller and callee
+are in the same assembly -- any use will generate a diagnostic and the caller is
+expected to suppress it, with the usual means (i.e. `#pragma` or project-wide
+`NoWarn`).
+
+## Tracking Items
+
+* The overall work of enabling this feature for libraries
+    - [dotnet/runtime#77869](https://github.com/dotnet/runtime/issues/77869)
+* Shipping `RequiresPreviewFeaturesAttribute` for downlevel
+    - [dotnet/runtime#79479](https://github.com/dotnet/runtime/issues/79479)
```

Comment:
Do we still need to ship `RequiresPreviewFeaturesAttribute` down-level, or would it be the new attribute that gets shipped down-level?

Reply:
I think only the new attribute is needed on down-level platforms.

Reply:
Yep